tutorials/02_functions-types.md

```diff
@@ -0,0 +1,518 @@
+
+```haskell
+type String = [Char]
+
+data Bool = True | False
+data Tree a = Leaf a | Branch (Tree a) (Tree a)
+            deriving (Show, Read)
+
+newtype Age = Age Int
+
+myTree :: Tree Int
+myTree = Branch (Leaf -7) (Branch (Leaf 7) (Leaf 10))
+
+myAge :: Age
+myAge = Age 20
+```
+
+The keyword `deriving` allows you to automatically make your data type instance of some typeclass (limited set of built-in classes) to allow some common Haskell mechanisms and functions to work with your types:
+
+* `Eq` - equality operators `==` and `/=`
+* `Ord` - comparison operators `<`, `<=`, `>`, `>=`; `min`, `max`, and `compare` (subclass of `Eq`)
+* `Show` - `show` for value to `String` conversion (+ other related functions)
+* `Read` - `read` for`String` to value conversion (+ other related functions)
+* `Enum` - for enumerations, allows the use of `..` range list syntax such as `[Blue .. Green]`
+* `Bounded` - for enumerations or other bounded, `minBound` and `maxBound` as the lowest and highest values that the type can take
+
+As it was said, typeclasses are very important in Haskell and will be covered later on. You will also learn how to make new typeclasses, their instances, etc.
+
+## Data types
+
+Haskell has a strong static type system which is one of the things making it so great. As we already saw, every expression in Haskell has some type and the type cannot change during runtime (that is the difference with dynamic typing). As in other programming languages can use predefined data types, get more from some libraries or introduce your own.
+
+### Basic Data Types
+
+* `Int` = A fixed-precision integer type with at least the range `[-2^29 .. 2^29-1]` (exact range can vary based on implementation, it can be check with `minBound` and `maxBound`)
+* `Integer` = Arbitrary-precision integers (e.g. theoretically unlimited)
+* `Float` = Single-precision floating point numbers
+* `Double` = Double-precision floating point numbers
+* `Word` = Unsigned integral number (same size as `Int`)
+* `Char` = Unicode character (ISO/IEC 10646)
+* `Bool` = truth value, only `True` or `False`
+* `String` = literally list of characters
+
+### Type and data constructor
+
+Get back to creating own data types with the `data` keyword. After `data` is the type construction starting with a capital letter and then there might be type parameters (type variables). Then `=` follows and after it, we can specify multiple data constructors separated by `|`. Each data constructor again starts with a capital letter and can be followed by data types which it takes as arguments.
+
+```haskell
+data MyType a b = MyTypeC1 a Int | MyTypeC2 String b | MyType3
+                deriving Show
+
+x :: MyType Bool Float
+x = MyTypeC1 True 10
+```
+
+Usually, when there is just one data constructor, the name is the same as of the type constructor (but it is not a rule).
```

Comment:
This is one another great Haskell feature: Algebraic Data Types (ADT). These are now found in more and more new languages (e.g. 'Rust' and very fresh 'Reason' from Facebook)